Sources/EmbraceCore/Capture/UX/View/UIViewControllerHandler.swift

```diff
@@ -0,0 +1,361 @@
+
+    init() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(appDidEnterBackground),
+            name: UIApplication.didEnterBackgroundNotification,
+            object: nil
+        )
+    }
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    func parentSpan(for vc: UIViewController) -> Span? {
+        guard let id = vc.emb_identifier else {
+            return nil
+        }
+
+        return parentSpans[id]
+    }
+
+    @objc func appDidEnterBackground() {
+        // end all parent spans and visibility spans if the app enters the background
+        // also clear all the cached spans
+        queue.async {
+            for span in self.visibilitySpans.values {
+                span.end()
+            }
+
+            for id in self.parentSpans.keys {
+                self.forcefullyEndSpans(id: id)
+            }
+
+            self.parentSpans.removeAll()
+            self.viewDidLoadSpans.removeAll()
+            self.viewWillAppearSpans.removeAll()
+            self.viewDidAppearSpans.removeAll()
+            self.visibilitySpans.removeAll()
+            self.uiReadySpans.removeAll()
+            self.alreadyFinishedUiReadyIds.removeAll()
+        }
+    }
+
+    func onViewDidLoadStart(_ vc: UIViewController) {
+
+        guard dataSource?.state == .active,
+              dataSource?.instrumentFirstRender == true,
+              vc.emb_shouldCaptureView,
+              let otel = dataSource?.otel else {
+            return
+        }
+
+        queue.async {
+            // generate id
+            let id = UUID().uuidString
+            vc.emb_identifier = id
+
+            // generate parent span
+            let className = vc.className
+
+            // check if with need to measure time-to-render or time-to-interactive
+            let nameFormat = vc is InteractableViewController ?
+                SpanSemantics.View.timeToInteractiveName :
+                SpanSemantics.View.timeToFirstRenderName
+
+            let spanName = nameFormat.replacingOccurrences(of: "NAME", with: className)
+
+            let parentSpan = self.createSpan(
+                with: otel,
+                vc: vc,
+                name: spanName
+            )
+
+            // generate view did load span
+            let viewDidLoadSpan = self.createSpan(
+                with: otel,
+                vc: vc,
+                name: SpanSemantics.View.viewDidLoadName,
+                parent: parentSpan
+            )
+
+            self.parentSpans[id] = parentSpan
+            self.viewDidLoadSpans[id] = viewDidLoadSpan
+        }
+    }
+
+    func onViewDidLoadEnd(_ vc: UIViewController) {
+        queue.async {
+            guard let id = vc.emb_identifier,
+                  let span = self.viewDidLoadSpans[id] else {
+                return
+            }
+
+            span.end()
+            self.viewDidLoadSpans[id] = nil
```

Comment:
Nitpick: to prevent from accessing twice the dictionary (which also has the locking mechanism) maybe you can do something like:
```suggestion
            guard let id = vc.emb_identifier,
                  let span = self.viewDidLoadSpans.removeValue(forKey: id) else {
                return
            }

            span.end()
```